# Pandas Essentials (Series, DataFrame, I/O, Cleaning)

In [2]:
import pandas as pd
import numpy as np

## 1) Φόρτωμα & Επιθεώρηση

In [8]:
df = pd.read_csv("../../data/data_raw/creditcard.csv")

df.head()         # πρώτες 5 γραμμές

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
0,0.0,-1.359807,-0.072781,2.536347,1.378155,-0.338321,0.462388,0.239599,0.098698,0.363787,...,-0.018307,0.277838,-0.110474,0.066928,0.128539,-0.189115,0.133558,-0.021053,149.62,0
1,0.0,1.191857,0.266151,0.166480,0.448154,0.060018,-0.082361,-0.078803,0.085102,-0.255425,...,-0.225775,-0.638672,0.101288,-0.339846,0.167170,0.125895,-0.008983,0.014724,2.69,0
2,1.0,-1.358354,-1.340163,1.773209,0.379780,-0.503198,1.800499,0.791461,0.247676,-1.514654,...,0.247998,0.771679,0.909412,-0.689281,-0.327642,-0.139097,-0.055353,-0.059752,378.66,0
3,1.0,-0.966272,-0.185226,1.792993,-0.863291,-0.010309,1.247203,0.237609,0.377436,-1.387024,...,-0.108300,0.005274,-0.190321,-1.175575,0.647376,-0.221929,0.062723,0.061458,123.50,0
4,2.0,-1.158233,0.877737,1.548718,0.403034,-0.407193,0.095921,0.592941,-0.270533,0.817739,...,-0.009431,0.798278,-0.137458,0.141267,-0.206010,0.502292,0.219422,0.215153,69.99,0


Μετατροπές τύπων

In [9]:
df.tail()         # τελευταίες 5 γραμμές

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
284802,172786.0,-11.881118,10.071785,-9.834783,-2.066656,-5.364473,-2.606837,-4.918215,7.305334,1.914428,...,0.213454,0.111864,1.014480,-0.509348,1.436807,0.250034,0.943651,0.823731,0.77,0
284803,172787.0,-0.732789,-0.055080,2.035030,-0.738589,0.868229,1.058415,0.024330,0.294869,0.584800,...,0.214205,0.924384,0.012463,-1.016226,-0.606624,-0.395255,0.068472,-0.053527,24.79,0
284804,172788.0,1.919565,-0.301254,-3.249640,-0.557828,2.630515,3.031260,-0.296827,0.708417,0.432454,...,0.232045,0.578229,-0.037501,0.640134,0.265745,-0.087371,0.004455,-0.026561,67.88,0
284805,172788.0,-0.240440,0.530483,0.702510,0.689799,-0.377961,0.623708,-0.686180,0.679145,0.392087,...,0.265245,0.800049,-0.163298,0.123205,-0.569159,0.546668,0.108821,0.104533,10.00,0
284806,172792.0,-0.533413,-0.189733,0.703337,-0.506271,-0.012546,-0.649617,1.577006,-0.414650,0.486180,...,0.261057,0.643078,0.376777,0.008797,-0.473649,-0.818267,-0.002415,0.013649,217.00,0


In [10]:
df.sample(5)      # τυχαίο δείγμα

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
39791,39968.0,1.162366,-0.248152,0.947749,-0.019912,-1.051075,-0.763135,-0.414792,-0.117607,0.505680,...,0.012199,-0.009256,0.035248,0.482606,0.044694,0.951183,-0.052650,0.027331,52.33,0
277120,167475.0,0.191534,0.759391,0.161786,-0.780663,0.873322,-0.703874,1.183021,-0.371947,0.177859,...,-0.360828,-0.686876,-0.056838,-0.666234,-0.430414,0.153353,0.082175,-0.133189,1.98,0
185570,126726.0,1.573559,-0.819552,-0.590449,1.077063,0.008474,1.516697,-0.770729,0.601971,0.856884,...,0.395931,1.042420,0.085312,-1.463117,-0.393187,-0.468667,0.074710,-0.037308,124.50,0
206686,136319.0,0.011693,0.611036,0.825998,-0.065768,0.506072,-0.663256,0.919985,-0.257695,-0.373298,...,-0.178510,-0.336115,0.006626,0.061396,-0.906471,0.244474,-0.035410,-0.049110,0.89,0
139907,83421.0,-1.440920,0.888603,-0.260411,-0.630516,-1.237166,-0.939017,0.901957,0.563055,-0.208439,...,0.089074,0.169221,0.250662,0.437092,-0.443669,0.835274,0.031644,-0.027815,180.14,0


In [11]:
df.shape          # (rows, cols)

(284807, 31)

In [12]:
df.info()         # πληροφορίες στηλών

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [13]:
df.dtypes         # τύποι


Time      float64
V1        float64
V2        float64
V3        float64
V4        float64
V5        float64
V6        float64
V7        float64
V8        float64
V9        float64
V10       float64
V11       float64
V12       float64
V13       float64
V14       float64
V15       float64
V16       float64
V17       float64
V18       float64
V19       float64
V20       float64
V21       float64
V22       float64
V23       float64
V24       float64
V25       float64
V26       float64
V27       float64
V28       float64
Amount    float64
Class       int64
dtype: object

In [14]:
df.describe()     # αριθμητικά στατιστικά


,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
count,284807.000000,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,...,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,284807.000000,284807.000000
mean,94813.859575,1.175161e-15,3.384974e-16,-1.379537e-15,2.094852e-15,1.021879e-15,1.494498e-15,-5.620335e-16,1.149614e-16,-2.414189e-15,...,1.628620e-16,-3.576577e-16,2.618565e-16,4.473914e-15,5.109395e-16,1.686100e-15,-3.661401e-16,-1.227452e-16,88.349619,0.001727
std,47488.145955,1.958696e+00,1.651309e+00,1.516255e+00,1.415869e+00,1.380247e+00,1.332271e+00,1.237094e+00,1.194353e+00,1.098632e+00,...,7.345240e-01,7.257016e-01,6.244603e-01,6.056471e-01,5.212781e-01,4.822270e-01,4.036325e-01,3.300833e-01,250.120109,0.041527
min,0.000000,-5.640751e+01,-7.271573e+01,-4.832559e+01,-5.683171e+00,-1.137433e+02,-2.616051e+01,-4.355724e+01,-7.321672e+01,-1.343407e+01,...,-3.483038e+01,-1.093314e+01,-4.480774e+01,-2.836627e+00,-1.029540e+01,-2.604551e+00,-2.256568e+01,-1.543008e+01,0.000000,0.000000
25%,54201.500000,-9.203734e-01,-5.985499e-01,-8.903648e-01,-8.486401e-01,-6.915971e-01,-7.682956e-01,-5.540759e-01,-2.086297e-01,-6.430976e-01,...,-2.283949e-01,-5.423504e-01,-1.618463e-01,-3.545861e-01,-3.171451e-01,-3.269839e-01,-7.083953e-02,-5.295979e-02,5.600000,0.000000
50%,84692.000000,1.810880e-02,6.548556e-02,1.798463e-01,-1.984653e-02,-5.433583e-02,-2.741871e-01,4.010308e-02,2.235804e-02,-5.142873e-02,...,-2.945017e-02,6.781943e-03,-1.119293e-02,4.097606e-02,1.659350e-02,-5.213911e-02,1.342146e-03,1.124383e-02,22.000000,0.000000
75%,139320.500000,1.315642e+00,8.037239e-01,1.027196e+00,7.433413e-01,6.119264e-01,3.985649e-01,5.704361e-01,3.273459e-01,5.971390e-01,...,1.863772e-01,5.285536e-01,1.476421e-01,4.395266e-01,3.507156e-01,2.409522e-01,9.104512e-02,7.827995e-02,77.165000,0.000000
max,172792.000000,2.454930e+00,2.205773e+01,9.382558e+00,1.687534e+01,3.480167e+01,7.330163e+01,1.205895e+02,2.000721e+01,1.559499e+01,...,2.720284e+01,1.050309e+01,2.252841e+01,4.584549e+00,7.519589e+00,3.517346e+00,3.161220e+01,3.384781e+01,25691.160000,1.000000


In [15]:
df.describe(include="all")  # όλα τα είδη στηλών

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
count,284807.000000,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,...,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,284807.000000,284807.000000
mean,94813.859575,1.175161e-15,3.384974e-16,-1.379537e-15,2.094852e-15,1.021879e-15,1.494498e-15,-5.620335e-16,1.149614e-16,-2.414189e-15,...,1.628620e-16,-3.576577e-16,2.618565e-16,4.473914e-15,5.109395e-16,1.686100e-15,-3.661401e-16,-1.227452e-16,88.349619,0.001727
std,47488.145955,1.958696e+00,1.651309e+00,1.516255e+00,1.415869e+00,1.380247e+00,1.332271e+00,1.237094e+00,1.194353e+00,1.098632e+00,...,7.345240e-01,7.257016e-01,6.244603e-01,6.056471e-01,5.212781e-01,4.822270e-01,4.036325e-01,3.300833e-01,250.120109,0.041527
min,0.000000,-5.640751e+01,-7.271573e+01,-4.832559e+01,-5.683171e+00,-1.137433e+02,-2.616051e+01,-4.355724e+01,-7.321672e+01,-1.343407e+01,...,-3.483038e+01,-1.093314e+01,-4.480774e+01,-2.836627e+00,-1.029540e+01,-2.604551e+00,-2.256568e+01,-1.543008e+01,0.000000,0.000000
25%,54201.500000,-9.203734e-01,-5.985499e-01,-8.903648e-01,-8.486401e-01,-6.915971e-01,-7.682956e-01,-5.540759e-01,-2.086297e-01,-6.430976e-01,...,-2.283949e-01,-5.423504e-01,-1.618463e-01,-3.545861e-01,-3.171451e-01,-3.269839e-01,-7.083953e-02,-5.295979e-02,5.600000,0.000000
50%,84692.000000,1.810880e-02,6.548556e-02,1.798463e-01,-1.984653e-02,-5.433583e-02,-2.741871e-01,4.010308e-02,2.235804e-02,-5.142873e-02,...,-2.945017e-02,6.781943e-03,-1.119293e-02,4.097606e-02,1.659350e-02,-5.213911e-02,1.342146e-03,1.124383e-02,22.000000,0.000000
75%,139320.500000,1.315642e+00,8.037239e-01,1.027196e+00,7.433413e-01,6.119264e-01,3.985649e-01,5.704361e-01,3.273459e-01,5.971390e-01,...,1.863772e-01,5.285536e-01,1.476421e-01,4.395266e-01,3.507156e-01,2.409522e-01,9.104512e-02,7.827995e-02,77.165000,0.000000
max,172792.000000,2.454930e+00,2.205773e+01,9.382558e+00,1.687534e+01,3.480167e+01,7.330163e+01,1.205895e+02,2.000721e+01,1.559499e+01,...,2.720284e+01,1.050309e+01,2.252841e+01,4.584549e+00,7.519589e+00,3.517346e+00,3.161220e+01,3.384781e+01,25691.160000,1.000000


In [80]:
print(df.columns.tolist())
df["Amount"] = df["Amount"].astype(float)


['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount', 'Class']


## 2) Επιλογές / Φιλτράρισμα / Νέες Στήλες (≈45’)

Επιλογές

In [81]:
# μία στήλη (Series)
df["Amount"]


0         149.62
1           2.69
2         378.66
3         123.50
4          69.99
           ...  
284802      0.77
284803     24.79
284804     67.88
284805     10.00
284806    217.00
Name: Amount, Length: 284807, dtype: float64

In [82]:
# πολλές στήλες (DataFrame)
df[["Amount", "Class"]]     # π.χ. Amount και Class

,Amount,Class
0,149.62,0
1,2.69,0
2,378.66,0
3,123.50,0
4,69.99,0
...,...,...
284802,0.77,0
284803,24.79,0
284804,67.88,0
284805,10.00,0


In [83]:
# με labels (loc)
df.loc[0:5, ["Amount", "Class"]]

,Amount,Class
0,149.62,0
1,2.69,0
2,378.66,0
3,123.50,0
4,69.99,0
5,3.67,0


In [84]:
# με θέσεις (iloc)
df.iloc[0:5, 0:2]   # τις πρώτες 2 στήλες για τις γραμμές 0–5

,Time,V1
0,0.0,-1.359807
1,0.0,1.191857
2,1.0,-1.358354
3,1.0,-0.966272
4,2.0,-1.158233


Φιλτράρισμα

In [85]:
df[df["Amount"] > 1000]

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
51,36.0,-1.004929,-0.985978,-0.038039,3.710061,-6.631951,5.122103,4.371691,-2.006868,-0.278736,...,1.393406,-0.381671,0.969719,0.019445,0.570923,0.333278,0.857373,-0.075538,1402.95,0
89,59.0,-0.773293,-4.146007,-0.932038,0.027094,-1.698307,0.460188,0.737344,-0.314216,-0.842673,...,0.890978,0.026123,-1.134769,-0.654958,0.098386,-0.209150,-0.171709,0.208057,1142.02,0
164,103.0,-6.093248,-12.114213,-5.694973,3.294389,-1.413792,4.776000,4.808426,-0.228197,-0.525896,...,2.228823,-2.264037,-3.553381,1.215279,-0.406073,-0.653384,-0.711222,0.672900,3828.04,0
633,481.0,-2.752124,-3.232168,0.873036,0.108217,-2.451279,0.486159,3.152907,-1.013087,1.128326,...,-0.299791,-0.003767,0.545794,0.853981,-0.943651,0.607826,0.335131,-0.561721,1015.61,0
865,656.0,-1.789835,-5.835768,-1.472791,-0.851160,-2.433875,0.401415,1.354543,-0.288283,2.426889,...,1.081404,-0.178233,-1.635914,-0.645651,0.112930,-0.726724,-0.184806,0.296762,1602.99,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
283875,171943.0,0.087032,-2.727628,-3.215031,2.216197,-0.229633,-1.419572,2.553476,-0.987507,-0.614069,...,0.969833,0.738640,-1.082522,0.535977,0.474597,-0.330842,-0.235690,0.105630,1026.46,0
284154,172199.0,-1.448606,-2.673922,-3.563343,-0.194096,0.007561,4.999178,4.136303,0.159725,-2.951736,...,0.644819,0.520282,2.202927,0.683422,0.446953,1.403290,-0.242420,0.221515,1088.04,0
284249,172273.0,-9.030538,-11.112584,-16.233798,3.592021,-40.427726,23.917837,44.054461,-7.277778,-4.210637,...,-0.269048,0.988144,7.040028,0.347693,2.520869,2.342495,3.478175,-2.713136,10199.44,0
284383,172397.0,-0.065120,-4.114668,-2.106854,-0.137339,-1.384530,0.239999,0.893833,-0.257922,1.145888,...,0.395868,-1.244969,-0.686403,-0.963878,-0.924685,0.210599,-0.253408,0.125687,1114.00,0


In [86]:
# Transactions με Amount > 1000 και Class == 1 (δηλαδή fraud)
df[(df["Amount"] > 1000) & (df["Class"] == 1)]


,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
6971,9064.0,-3.499108,0.258555,-4.489558,4.853894,-6.974522,3.628382,5.431271,-1.946734,-0.775680,...,-1.052368,0.204817,-2.119007,0.170279,-0.393844,0.296367,1.985913,-0.900452,1809.68,1
10690,18088.0,-12.224021,3.854150,-12.466766,9.648311,-2.726961,-4.445610,-21.922811,0.320792,-4.433162,...,-1.159830,-1.504119,-19.254328,0.544867,-4.781606,-0.007772,3.052358,-0.775036,1218.89,1
81609,59011.0,-2.326922,-3.348439,-3.513408,3.175060,-2.815137,-0.203363,-0.892144,0.333226,-0.802005,...,1.226648,-0.695902,-1.478490,-0.061553,0.236155,0.531911,0.302324,0.536375,1389.56,1
89190,62467.0,-5.344665,-0.285760,-3.835616,5.337048,-7.609909,3.874668,1.289630,0.201742,-3.003532,...,0.276011,1.342045,-1.016579,-0.071361,-0.335869,0.441044,1.520613,-1.115937,1402.16,1
95597,65385.0,-2.923827,1.524837,-3.018758,3.289291,-5.755542,2.218276,-0.509995,-3.569444,-1.016592,...,-0.511657,-0.122724,-4.288639,0.563797,-0.949451,-0.204532,1.510206,-0.324706,1354.25,1
176049,122608.0,-2.003460,-7.159042,-4.050976,1.309580,-2.058102,-0.098621,2.880083,-0.727484,1.460381,...,1.244287,-1.015232,-1.800985,0.657586,-0.435617,-0.894509,-0.397557,0.314262,2125.87,1
199896,133184.0,-1.212682,-2.484824,-6.397186,3.670562,-0.863375,-1.855855,1.017732,-0.544704,-1.703378,...,1.396872,0.092073,-1.492882,-0.204227,0.532511,-0.293871,0.212663,0.431095,1335.00,1
249167,154278.0,-1.600211,-3.488130,-6.459303,3.246816,-1.614608,-1.260375,0.288223,-0.048964,-0.734975,...,1.191175,-0.967141,-1.463421,-0.624231,-0.176462,0.400348,0.152947,0.477775,1504.93,1
249239,154309.0,-0.082983,-3.935919,-2.616709,0.163310,-1.400952,-0.809419,1.501580,-0.471000,1.519743,...,0.702672,-0.182305,-0.921017,0.111635,-0.071622,-1.125881,-0.170947,0.126221,1096.99,1


In [87]:
# Transactions μεταξύ 500 και 1000 ευρώ
df[(df["Amount"] >= 500) & (df["Amount"] <= 1000)]


,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
134,83.0,-1.897331,0.955626,0.052543,1.276656,-3.323084,3.229911,1.029631,1.515607,-0.059627,...,0.128416,0.776078,0.477537,-0.608981,-1.120892,-0.413851,0.061399,-0.187964,552.18,0
140,87.0,-5.101877,1.897022,-3.458034,-1.277543,-5.517758,2.098366,3.329603,1.250966,0.271501,...,-0.871744,-0.678879,-0.555900,-0.761660,0.066611,0.767227,0.731634,-0.860310,919.60,0
150,93.0,-1.640014,-0.479831,1.613630,2.997979,-3.803788,5.051812,3.315254,-0.303785,0.671395,...,-0.367933,0.641891,-0.405779,-1.268700,0.494090,0.610967,0.608993,-0.410034,937.69,0
174,112.0,-4.816022,-3.345061,0.566006,3.960568,-2.495400,3.691365,1.603360,-0.071801,1.092418,...,-0.797155,0.260093,-2.712276,-1.020436,0.070990,0.619086,1.720707,-1.648553,937.69,0
227,152.0,0.095490,-2.557694,-1.134055,-0.435770,0.728493,4.090476,-0.429329,0.896879,1.004244,...,0.080098,-1.038538,-0.489645,1.108440,0.072603,0.893183,-0.146017,0.113110,614.87,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
284542,172543.0,-0.909722,0.577334,-0.610814,3.223744,-2.216835,2.656914,3.166607,0.099503,-1.175961,...,0.287250,1.039311,0.286808,0.531633,0.263987,0.221902,0.053983,-0.229425,707.21,0
284573,172569.0,-1.762664,-1.230888,-0.587727,-1.898574,-2.611232,1.301447,3.645126,-0.911945,1.133300,...,-0.526145,-0.121367,0.631407,-0.416585,-1.338263,0.006724,-0.084768,0.308717,724.00,0
284703,172699.0,1.019594,-1.956473,-1.431268,0.539727,-0.499995,0.301251,0.326079,-0.062475,1.121288,...,-0.081894,-1.202749,-0.182683,-1.075679,-0.417648,-1.080842,-0.049144,0.039078,521.90,0
284712,172704.0,0.470194,-1.619056,-2.258427,4.440719,0.202024,-0.458243,1.970565,-0.665303,-1.630737,...,0.603016,0.104952,-0.716836,0.121688,0.170587,0.082567,-0.218942,0.070955,787.95,0


Νέες στήλες

In [88]:
df["Amount_eur"] = df["Amount"] * 0.92
print(df)


            Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V22       V23  \
0       0.462388  0.239599  0.098698  0.363787  ...  0.27783

Missing values

In [89]:
# Συνολικός έλεγχος
print(df.isna().sum().sum())   # όλα πρέπει να βγουν 0



0


In [90]:
# ή αναλυτικά
print(df.isna().sum())

Time          0
V1            0
V2            0
V3            0
V4            0
V5            0
V6            0
V7            0
V8            0
V9            0
V10           0
V11           0
V12           0
V13           0
V14           0
V15           0
V16           0
V17           0
V18           0
V19           0
V20           0
V21           0
V22           0
V23           0
V24           0
V25           0
V26           0
V27           0
V28           0
Amount        0
Class         0
Amount_eur    0
dtype: int64


In [91]:
print(df.columns)

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class', 'Amount_eur'],
      dtype='object')


In [92]:
# drop συγκεκριμένες στήλες
df = df.drop(columns=["Amount_eur"])

# έλεγχος
print(df.columns)

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')


## 3) Ομαδοποιήσεις / Στατιστικά / Ταξινομήσεις

- Ομαδοποίηση & συναθροίσεις

In [93]:
df.groupby("Class")["Amount"].agg(["count", "mean", "sum"])

,count,mean,sum
Class,,,
0,284315,88.291022,25102462.04
1,492,122.211321,60127.97


- Κατανομή κατηγορίας (ποσοστά)

In [94]:
df["Class"].value_counts(normalize=True) # ποσοστά ανά κατηγορία 

Class
0    0.998273
1    0.001727
Name: proportion, dtype: float64

- Μεγαλύτερα ποσά – 2 ισοδύναμοι δρόμοι

In [95]:
df.sort_values("Amount", ascending=False).head(5)  # top 5 γραμμές με βάση amount, επιστρέφει ολόκληρες γραμμές (DataFrame)


,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
274771,166198.0,-35.548539,-31.850484,-48.325589,15.304184,-113.743307,73.301626,120.589494,-27.347360,-3.872425,...,-21.620120,5.712303,-1.581098,4.584549,4.554683,3.415636,31.612198,-15.430084,25691.16,0
58465,48401.0,-36.802320,-63.344698,-20.645794,16.715537,-20.672064,7.694002,24.956587,-4.730111,-2.687312,...,11.455313,-10.933144,-17.173665,1.180700,-7.025783,-2.534330,-3.602479,3.450224,19656.53,0
151296,95286.0,-34.549296,-60.464618,-21.340854,16.875344,-19.229075,6.335259,24.422716,-4.964566,0.188912,...,11.502580,-9.499423,-16.513186,0.744341,-7.081325,-2.604551,-3.550963,3.250802,18910.00,0
46841,42951.0,-23.712839,-42.172688,-13.320825,9.925019,-13.945538,5.564891,15.710644,-2.844253,-1.580725,...,7.921600,-6.320710,-11.310338,0.404175,-4.547278,-1.577118,-2.357385,2.253662,12910.93,0
54018,46253.0,-21.780665,-38.305310,-12.122469,9.752791,-12.880794,4.256017,14.785051,-2.818253,-0.667338,...,7.437478,-5.619439,-10.547038,0.653249,-4.232409,-0.480459,-2.257913,2.082488,11898.09,0


In [96]:
df["Amount"].nlargest(5) # top 5 τιμές της στήλης amount, επιστρέφει μόνο τις τιμές (Series). nlargest είναι συνήθως πιο γρήγορο και βολικό όταν σε νοιάζουν μόνο οι τιμές (ή μαζί με index).

274771    25691.16
58465     19656.53
151296    18910.00
46841     12910.93
54018     11898.09
Name: Amount, dtype: float64

## 4) Joins / Merge

In [ ]:
# μικρό lookup για να δώσουμε ονόματα στις κλάσεις
df_classes = pd.DataFrame({ # dictionary
    "Class": [0, 1], # οι τιμές που υπάρχουν στην αρχική στήλη
    "ClassLabel": ["Legit", "Fraud"] # τα ονόματα που θέλουμε να δώσουμε
})

merged = pd.merge(df, df_classes, on="Class", how="left") # αριστερό join
merged[["Amount", "Class", "ClassLabel"]].head() # έλεγχος



,Amount,Class,ClassLabel
0,149.62,0,Legit
1,2.69,0,Legit
2,378.66,0,Legit
3,123.50,0,Legit
4,69.99,0,Legit


5) Mini-Ασκήσεις

5.1 Μέτρημα ποσοστού απάτης “class 1” (fraud ratio) στο dataset

In [101]:
# 5.1 Fraud ratio (Class=1)
fraud_ratio = df["Class"].value_counts(normalize=True).get(1, 0) # normalize=True: επιστρέφει ποσοστά αντί για counts.

print(f"Fraud ratio: {fraud_ratio:.4%}") # Το print με :.4% το δείχνει ως ποσοστό με 4 δεκαδικά (π.χ. 0.001728 → 0.1728%).

Fraud ratio: 0.1727%


5.2 Median Amount ανά κλάση (Legit vs Fraud)

- Αν ο median των Fraud είναι σημαντικά μεγαλύτερος ή μικρότερος από τον median των Legit, αυτό είναι διαγνωστικό χαρακτηριστικό → μπορεί να βοηθήσει σε feature engineering (π.χ. transaction amount thresholds).
- Αν είναι παρόμοιοι, τότε το Amount μόνο του δεν είναι καλό διαχωριστικό, και χρειάζονται συνδυασμοί με άλλα features.

In [ ]:
df.groupby("Class")["Amount"].median()

Class
0    22.00
1     9.25
Name: Amount, dtype: float64

5.3 Mean amount ανά ώρα ημέρας

In [103]:
# Δημιουργία στήλης ώρας ημέρας (0–23)
df["hour_of_day"] = (df["Time"] // 3600) % 24  

# Μέσο ποσό ανά ώρα
mean_amount_by_hour = df.groupby("hour_of_day")["Amount"].mean()
print(mean_amount_by_hour)


hour_of_day
0.0      60.325960
1.0      62.636363
2.0      70.258618
3.0      52.076793
4.0      77.039036
5.0      50.675645
6.0      65.060278
7.0      67.818459
8.0      88.771243
9.0     102.843151
10.0    109.673116
11.0    109.613287
12.0    106.045513
13.0     98.959359
14.0    103.538500
15.0    101.999247
16.0    103.483250
17.0     98.323639
18.0     77.252852
19.0     77.986475
20.0     74.544826
21.0     71.661028
22.0     69.876031
23.0     68.913253
Name: Amount, dtype: float64


🔎 Παρατηρήσεις

- Νύχτα (0:00–5:00) → τα amounts είναι χαμηλά (≈ 50–77€).
Αυτό ταιριάζει με το ότι εκείνες τις ώρες γίνονται λίγες συναλλαγές και συνήθως μικρές (π.χ. online πληρωμές, μικρές αγορές).
- Πρωί–Μεσημέρι (8:00–13:00) → βλέπουμε σταθερή άνοδο (≈ 88–110€).
Το peak είναι γύρω στις 10–11 το πρωί (≈ 109€). Εκείνη την ώρα μάλλον γίνονται πιο "σοβαρές" συναλλαγές (π.χ. επαγγελματικές).
- Απόγευμα–Βράδυ (14:00–20:00) → οι τιμές μένουν γύρω στα 100€ μέχρι τις 16:00 και μετά πέφτουν (≈ 74–77€).
- Αργά το βράδυ (21:00–23:00) → ξαναπέφτει στα ~69€.

✅ Συμπέρασμα

- Οι μεγαλύτερες συναλλαγές γίνονται πρωινές ώρες (9–12), με peak γύρω στις 10:00–11:00.
- Οι χαμηλότερες μέσες αξίες γίνονται νωρίς το πρωί/ξημερώματα (3:00–6:00).
- Αυτό μπορεί να αξιοποιηθεί στο fraud analysis:
- Αν δεις πολύ μεγάλες συναλλαγές στις 3–4 το πρωί, ίσως είναι πιο ύποπτες.
- Αν δεις πολύ μικρές συναλλαγές το πρωί (10–11), ίσως είναι outlier σε σχέση με τη συνήθη συμπεριφορά.